Sprint 7 solution in file manager

fzk273 opened the pull request.

Commit:
add load from file functionality
add Enum with task types
add Save from file
add custom exception
add Tests for FBTM class
add new Readme

Ivanov-s-tmn commented on src/ru/prakticum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,163 @@
+                String taskName = splittedLine.get(2);
+                Status taskStatus = Status.valueOf(splittedLine.get(3));
+                String taskDescription = splittedLine.get(4);
+                //TODO я практически уверен, что всё это можно реализовать сильно лаконичнее. need help
+                if (taskType.equals(TaskType.TASK)) {
+                    Task taskFromBackUP = new Task(taskName, taskDescription);
+                    taskFromBackUP.setId(taskId);
+                    taskFromBackUP.setStatus(taskStatus);
+                    fileBackedTaskManager.tasks.put(taskId, taskFromBackUP);
+                }
+                if (taskType.equals(TaskType.EPIC)) {
+                    Epic epicFromBackUP = new Epic(taskName, taskDescription);
+                    epicFromBackUP.setId(taskId);
+                    epicFromBackUP.setStatus(taskStatus);
+                    fileBackedTaskManager.epics.put(taskId, epicFromBackUP);
+                }
+                if (taskType.equals(TaskType.SUBTASK)) {
+                    SubTask subtaskFromBackUP = new SubTask(taskName, taskDescription, Integer.valueOf(splittedLine.get(5)));
+                    subtaskFromBackUP.setId(taskId);
+                    subtaskFromBackUP.setStatus(taskStatus);
+                    fileBackedTaskManager.subtasks.put(taskId, subtaskFromBackUP);
+                }
```
Вполне неплохо реализовано)
Что можно сделать, это вынести парсинг строки и создание объекта задачи/подзадачи/эпика из строки в отдельный утилитный класс, получится что-то вроде:
`Task task = CSVTaskFormat.taskFromString(splittedLine);`
После этого уже добавляем полученные задачи в хешмапы, в зависимости от типа, можно тоже вынести в отдельный метод `fileBackedTaskManager.addTask(task);`
Единственный момент, который у тебя не учтен, это сабтаски нужно добавлять в список внутри эпика при выгрузке из файла

fzk273 replied:
отрефакторил эту часть.

Ivanov-s-tmn commented on src/ru/prakticum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,163 @@
+                line = reader.readLine();
+            }
+
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
`FileNotFoundException` отдельно отлавливать не нужно, это исключение-наследник `IOException`, выбрасывай тут тоже свое исключение `ManagerSaveException`

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+
+    public void save() {
```
Этот метод стоит закрыть модификатором доступа, он используется в логике переопределенных методов менеджера и не нужен для использования в качестве отдельного метода где-то вне этого класса

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/managers/InMemoryTaskManager.java:
```diff
@@ -27,6 +27,10 @@ public InMemoryTaskManager() {
         history = Managers.getDefaultHistory();
     }
 
+    public Integer getCounter() {
+        return counter;
+    }
```
Не нужно предоставлять публичный геттер для счетчика, так мы получаем ссылку на него и этот счетчик можно изменить из любого класса. Благодаря protected модификатор, он будет унаследован файловым менеджером, этого достаточно

fzk273 replied:
поправил

Ivanov-s-tmn commented on src/ru/prakticum/tasks/Epic.java:
```diff
@@ -1,15 +1,23 @@
 
 public class Epic extends Task {
     private ArrayList<Integer> subtaskIds;
+    private TaskType taskType;
```
Как таковое поле не нужно, достаточно одного геттера, который будет возвращать тип задачи, как пример для эпика:
```java
@Override
public TaskType getTaskType() {
     return TaskType.EPIC;
}
```

fzk273 replied:
поправил

Ivanov-s-tmn commented on test/ru/prakticum/managers/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+    private Task task;
+    private Epic epic;
+    private SubTask subTask;
+    private FileReader fileReader;
```
Для тестов не нужно использовать `FileReader`, нужно проверить сохранение данных в файл и восстановление менеджера из файла, примерный тест-кейс:
1. Создать файловый менеджер
2. Добавить, например по одной задаче каждого типа
3. Создать менеджер через `loadFromFile()` из этого же файла
4. Проверить, что списки задач/подзадач/эпиков менеджера до выгрузки и после выгрузки идентичны.

Такого одного теста будет достаточно для файлового менеджера. С помощью него мы проверим и сохранение в файл и чтение из файла

fzk273 replied:
поправил

Ivanov-s-tmn commented on test/ru/prakticum/managers/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+            Assertions.assertTrue(allLines.contains("5,EPIC,Epic name,NEW,Epic desc,"));
+            Assertions.assertTrue(allLines.contains("6,SUBTASK,SubTask name,NEW,SubTaskDesc,2"));
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Обработка исключений в тестах не лучшая практика

fzk273 replied:
переписал тесты, более не релевантно

Commit:
Code refactor after review
fix tests for FBTM
add static class for parsing string from file
fix task type getters
fix loadFromFile function
